frontends/precinct-scanner/src/components/replace_ballot_bag_modal.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { useEffect, useState } from 'react';
+import { Button, Text, Modal } from '@votingworks/ui';
+import { throwIllegalValue } from '@votingworks/utils';
+import { ScannedBallotCount } from './scanned_ballot_count';
+import { CenteredLargeProse } from './layout';
+import { BALLOT_BAG_CAPACITY } from '../config/globals';
+
+interface Props {
+  scannedBallotCount: number;
+  pollWorkerAuthenticated: boolean;
+  onComplete: VoidFunction;
+}
+
+type ModalState = 'no_auth' | 'confirming' | 'confirmed';
+
+export function ReplaceBallotBagModal({
+  scannedBallotCount,
+  pollWorkerAuthenticated,
+  onComplete,
+}: Props): JSX.Element {
+  const [modalState, setModalState] = useState<ModalState>('no_auth');
+
+  useEffect(() => {
+    if (modalState === 'no_auth' && pollWorkerAuthenticated) {
+      setModalState('confirming');
+    } else if (modalState === 'confirming' && !pollWorkerAuthenticated) {
+      setModalState('no_auth');
+    } else if (modalState === 'confirmed' && !pollWorkerAuthenticated) {
+      onComplete();
+    }
+  }, [modalState, onComplete, pollWorkerAuthenticated]);
```

Comment:
I think this is one of those cases where `useEffect` is not actually needed. I think in general it's good to try to reserve `useEffect` for side effects, not for managing UI state.

A sketch of how to mostly eliminate it:

```ts
const [confirmed, setConfirmed] = useState(false)

if (!confirmed && !pollWorkerAuthenticated) {
   // render ballot bag full message
}

if (!confirmed && pollWorkerAuthenticated) {
  // render confirmation screen
}

if (confirmed) {
  // render remove card screen
}
```
Then, the confirm button handler can be `setConfirmed(true)`.

You still need a `useEffect` to listen for the final card removal:
```ts
useEffect(() => {
  if (confirmed && !pollworkerAuthenticated)
    onComplete()
})
```

What do you think?

Reply:
Good call, thanks. I think the Effect was necessary when we were able to step back from the confirmation screen with the "cancel" button, but def not with the current flow.

Reply:
The only clause that might have to remain in the useEffect is - we have to trigger the callback when the poll worker card is removed & confirmed. That seems like a side effect to me?

Reply:
Yeah, generally the approach I try to use is to trigger state updates based on user actions (e.g. button press). That's easy when the user action is on screen. With hardware actions, you can't do it as easily due to the way we currently track hardware actions via a changing value. So using a `useEffect` there is a good workaround.

That being said, I'm noodling on how we might design a hardware interface that could let us register action handlers similar to on-screen actions. Let me know if you have ideas!

Reply:
Clearly a longer term convo but...

What if we had Listener Components, and hardware actions at the top level would trigger events on all listener components currently on the screen?

I don't know enough to know why that would be a better paradigm, the current paradigm seems acceptable to me.

Reply:
The risk with the current approach is that a `useEffect` might run at any time, not just on the change hardware status. Using the dependency array to get around this only sometimes works (e.g. it doesn't work when the dependency is an object that isn't memoized), so it's tricky to reason about. And React intends the dependency array to only be used for optimization, not correctness.

In contrast, with a button handler, you can trivially rely on it only getting called when the button gets pressed. Very easy to reason about!

I like your idea! I was thinking along similar lines!